// ts-mongoose/example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var createSchema_1 = require("./createSchema");
var Type_1 = require("./Type");
var AddressSchema = createSchema_1.createSchema({
    city: Type_1.Type.string(),
    country: Type_1.Type.optionalString(),
    zip: Type_1.Type.optionalString(),
});
// function blockUser(user: UserDoc) {}
function createAddress() {
    return {
        city: 'a',
    };
}
var PhoneSchema = createSchema_1.createSchema({ phoneNumber: Type_1.Type.number() }, { _id: false });
var exampleSchema = createSchema_1.createSchema({
    firstName: Type_1.Type.string(),
    lastName: Type_1.Type.string(),
    profile: Type_1.Type.array().of({
        picture: Type_1.Type.string(),
        phone: Type_1.Type.optionalString(),
    }),
    profile2: Type_1.Type.object().of({
        picture: Type_1.Type.string(),
        phone: Type_1.Type.optionalString(),
    }),
    address: Type_1.Type.optionalSchema().of(AddressSchema),
    phone: Type_1.Type.schema().of(PhoneSchema),
    phones: Type_1.Type.array().of(PhoneSchema),
});
var a = null;
